### PR TITLE
Implement Vulnerability windows

### PR DESCRIPTION
**Pull request by Aingar**
Also fix build modules saving/reading.

### CODE REVIEW

File: Plugins/Public/base_plugin/Main.cpp [hunk 1]
```diff
@@ -2337,7 +2356,18 @@ void __stdcall HkCb_AddDmgEntry(DamageList *dmg, unsigned short sID, float& newH
 			ConPrint(L"HkCb_AddDmgEntry[1] - invalid damage?\n");
 		return;
 	}
-
+	if (curr - newHealth > 1500000)
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/Main.cpp:2359]: I'm sure this was in another PR

File: Plugins/Public/base_plugin/Main.cpp [hunk 2]
```diff
@@ -2530,6 +2560,93 @@ bool ExecuteCommandString_Callback(CCmds* cmd, const wstring &args)
 		//cmd->Print(L"OK Base is gone are you proud of yourself.");
 		return true;
 	}
+	else if (args.find(L"basedespawn") == 0)
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/Main.cpp:2563]: This was definitely in another PR

**Review by dsyalex — CHANGES_REQUESTED** [review]
Requires rebase and also has overlap with other PRs

File: Plugins/Public/base_plugin/Main.cpp (hunk 2 above)
Comment (Aingar) [reply Plugins/Public/base_plugin/Main.cpp:2562]: Yep, admin commands, it'll dissapear from the diff once that one is merged.

File: Plugins/Public/base_plugin/BuildModule.cpp [hunk 3]
```diff
@@ -209,8 +209,7 @@ void BuildModule::LoadState(INI_Reader& ini)
 void BuildModule::SaveState(FILE* file)
 {
 	fprintf(file, "[BuildModule]\n");
-	fprintf(file, "build_type = %u\n", active_recipe.shortcut_number);
-	fprintf(file, "infotext = %s\n", wstos(active_recipe.infotext).c_str());
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/BuildModule.cpp:213]: what does removing infotext have to do with vulnerability windows?

File: Plugins/Public/base_plugin/CoreModule.cpp [hunk 4]
```diff
@@ -226,12 +226,20 @@ void CoreModule::SetShieldState(const int shieldState)
...
 	// Disable shield if time elapsed
-	if (base->shield_timeout < time)
+	if (base->shield_timeout && base->shield_timeout < time)
 	{
 		base->shield_timeout = 0;
 		base->shield_state = PlayerBase::SHIELD_STATE_ONLINE;
 		SetShieldState(base->shield_state);
 	}
+	
+	// we need to periodically set the health of all POBs to trigger a clientside 'refresh'
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/CoreModule.cpp:236]: I'm sure we did this in another PR - have you rebased?

File: Plugins/Public/base_plugin/Main.cpp [hunk 5]
```diff
@@ -1035,13 +1039,12 @@ void HkTimerCheckKick()
...
 		LoadSettingsActual();
 	}
-
 	uint curr_time = (uint)time(0);
-	isGlobalBaseInvulnerabilityActive = checkBaseVulnerabilityStatus();
 	for(auto& iter : player_bases)
 	{
 		PlayerBase *base = iter.second;
 		base->Timer(curr_time);
+		counter++;
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/Main.cpp]: what is this?

File: Plugins/Public/base_plugin/PlayerBase.cpp [hunk 6]
```diff
@@ -75,21 +73,72 @@ void PlayerBase::Spawn()
...
+		}
+		vulnerableWindowStatus = true;
+	}
+	else if (!single_vulnerability_window && IsVulnerabilityWindowActive(vulnerabilityWindow2, timeOfDay))
+	{
+		if (!vulnerableWindowStatus)
+		{
+			vulnerableWindowStatus = true;
+			siege_mode = true;
+			SyncReputationForBase();
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerBase.cpp:113]: what about all the other stuff we do above?

File: Plugins/Public/base_plugin/PlayerBase.cpp [hunk 7]
```diff
@@ -780,59 +849,64 @@ void PlayerBase::SpaceObjDamaged(uint space_obj, uint attacking_space_obj, float
...
+	}
+
+	// If the shield is not active but could be set a time 
+	// to request that it is activated.
+	if (!this->shield_timeout && this->isShieldOn == false
+		&& this->vulnerableWindowStatus)
+	{
+		const wstring& charname = (const wchar_t*)Players.GetActiveCharacterName(client);
+		ReportAttack(this->basename, charname, this->system);
+		if (set_plugin_debug > 1)
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerBase.cpp:909]: this doesn't look right

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 8]
```diff
@@ -2412,4 +2411,104 @@ namespace PlayerCommands
...
+		{
+			PrintUserCmdText(client, L"ERR Vulnerability windows can only be set to full hour values between 0 and 23");
+			return;
+		}
+
+		int vulnerabilityWindowOneStart = param1 * 60; // minutes
+		int vulnerabilityWindowOneEnd = (vulnerabilityWindowOneStart + vulnerability_window_length) % (60 * 24); // 
+
+		int vulnerabilityWindowTwoStart = param2 * 60;
+		int vulnerabilityWindowTwoEnd = vulnerabilityWindowTwoStart + vulnerability_window_length % (60 * 24);
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:2450]: add brackets to match window one

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 9]
```diff
@@ -2412,4 +2411,104 @@ namespace PlayerCommands
...
+
+		if (!pb)
+		{
+			PrintUserCmdText(client, L"ERR no base targeted");
+			return;
+		}
+
+		if (single_vulnerability_window)
+		{
+			PrintUserCmdText(client, L"This base has its vulnerability window starting at %u:00", pb->vulnerabilityWindow1.start / 60);
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:2507]: do we not also want to provide window ends?

File: Plugins/Public/base_plugin/PlayerBase.cpp (hunk 6 above)
Comment (Aingar) [reply Plugins/Public/base_plugin/PlayerBase.cpp:113]: Those are all separate else-if checks, its impossible more than 1 would execute

File: Plugins/Public/base_plugin/CoreModule.cpp (hunk 4 above)
Comment (Aingar) [reply Plugins/Public/base_plugin/CoreModule.cpp:236]: resolved

**Review by dsyalex — APPROVED** [review]

**Pull request merged**